streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -601,20 +601,34 @@ public final void connectProcessorAndStateStores(final String processorName,
     }
 
     public void connectSourceStoreAndTopic(final String sourceStoreName,
-                                            final String topic) {
+                                           final String topic) {
         if (storeToChangelogTopic.containsKey(sourceStoreName)) {
             throw new TopologyException("Source store " + sourceStoreName + " is already added.");
         }
         storeToChangelogTopic.put(sourceStoreName, topic);
     }
 
-    public final void addInternalTopic(final String topicName) {
+    public final void addInternalTopic(final String topicName,
+                                       final InternalTopicProperties internalTopicProperties) {
         Objects.requireNonNull(topicName, "topicName can't be null");
-        internalTopicNames.add(topicName);
+        Objects.requireNonNull(internalTopicProperties, "internalTopicProperties can't be null");
+
+        internalTopicNamesWithProperties.put(topicName, internalTopicProperties);
+    }
+
+    public final synchronized void copartitionSources(final Collection<String> sourceNodes) {
+        copartitionSourceGroups.add(new HashSet<>(sourceNodes));
     }
 
-    public final void copartitionSources(final Collection<String> sourceNodes) {
-        copartitionSourceGroups.add(Collections.unmodifiableSet(new HashSet<>(sourceNodes)));
+    public final synchronized void maybeUpdateCopartitionSourceGroups(final String oldNodeName,
```

Comment:
I'm not 100% sure if `synchronized` here  and in `copartitionSources(Collection)` is necessary. But `copartitionGroups()` method is synchronized, so I believe synchronizing other accessors to this collection makes sense.